src/editor/dag.rs

```diff
@@ -702,6 +703,60 @@ impl<'arena, Node: Ast<'arena>> Dag<'arena, Node> {
         self.write_text(&mut s, format);
         s
     }
+
+    ///  Generate `.dot` code of the internal state of the `Dag`
+    pub fn to_dot_code(&self) -> String {
+        /// Helper function that recursively builds dot code for the entire subtree of a given node
+        fn add_to_graph<'arena, Node: Ast<'arena>>(
+            node: &'arena Node,
+            digraph_edges: &mut String,
+            hmap_nodes: &mut HashMap<usize, String>,
+        ) {
+            let name = node as *const Node as usize;
+            hmap_nodes.entry(name).or_insert(format!(
+                "node{} [label=\"{}\"]\n",
+                name,
+                node.debug_name()
+            ));
```

Comment:
We're ending up with lines being added multiple times in the output `Dag`:

![dag](https://user-images.githubusercontent.com/32932209/106941727-222bb600-671b-11eb-997e-e1b309755723.png)

I think this would probably be fixed if we just returned out of the `add_to_graph` if we encounter a node that is already in the `HashMap`, like:

```rust
if hmap_nodes.entry(name).is_some() {
    return;
}

hmap_nodes.insert(format!(...));

....
```

Reply:
Yes. I wil

> This is looking cooooool!
> 
> I did some fiddling and I'm getting some very cool pictures:
> 
> ![dag](https://user-images.githubusercontent.com/32932209/106943036-ce21d100-671c-11eb-9246-23941be13ebb.png)
> 
> I think it would be good, though, if we force dot to put the node children in the order we gave them, as well as putting the roots in the right order... do you want to do that in this PR?

Yes. I'd like to fix this.

Reply:
> We're ending up with lines being added multiple times in the output `Dag`:
> 
> ![dag](https://user-images.githubusercontent.com/32932209/106941727-222bb600-671b-11eb-997e-e1b309755723.png)
> 
> I think this would probably be fixed if we just returned out of the `add_to_graph` if we encounter a node that is already in the `HashMap`, like:
> 
> ```rust
> if hmap_nodes.entry(name).is_some() {
>     return;
> }
> 
> hmap_nodes.insert(format!(...));
> 
> ....
> ```

I'm not sure the cause of this, feel like I added those edges twice somehow. I will fix this tonight.

Reply:
Can you tell me where did you call this `add_to_dot` function?